File: Linear/Dynamic/Linked List/Doubly Linked List/01_Traversing/Forward/Recursive/Traversing.py
class Node:
    def __init__(self, data):
        self.data = data
        self.prev = None
        self.next = None

# Recursive function for forward traversal
def forwardTraversal(head):
    if head is None:
        print()
        return
    
    # Print current node's data
    print(head.data, end="")
    if head.next is not None:
        print(" <-> ", end="")
    
    # Recursive call with the next node
    forwardTraversal(head.next)

# Create a hardcoded doubly linked list:
# 1 <-> 2 <-> 3
head = Node(1)
second = Node(2)
third = Node(3)

head.next = second
second.prev = head
second.next = third
third.prev = second

print("Forward Traversal: ", end="")
forwardTraversal(head)

File: Linear/Dynamic/Linked List/Doubly Linked List/01_Traversing/Forward/Recursive/Traversing.java
class Node {
    int data;
    Node prev;
    Node next;

    Node(int data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

class RecursiveForwardTraversing {
    // Recursive function for forward traversal
    public static void forwardTraversal(Node head) {
        if (head == null) {
            System.out.println();
            return;
        }

        // Print current node's data
        System.out.print(head.data);
        if (head.next != null)
        System.out.print(" <-> ");

        // Recursive call with the next node
        forwardTraversal(head.next);
    }

    public static void main(String[] args) {
        // Create a hardcoded doubly linked list:
        // 1 <-> 2 <-> 3
        Node head = new Node(1);
        Node second = new Node(2);
        Node third = new Node(3);

        head.next = second;
        second.prev = head;
        second.next = third;
        third.prev = second;

        System.out.print("Forward Traversal: ");
        forwardTraversal(head);
    }
}